### ISSUE TITLE
New 1.5 tag system issues

### ISSUE THREAD

**Issue by Tryibion** [FlaxEngine/FlaxEngine#885]
Hello,

I added a tag and a sub tag to that tag. closed the editor a couple of times and then the tags got messed it. I dont see the main tag `Testing` and the sub tag for it is inside of its check box. 

![tag issue](https://user-images.githubusercontent.com/71274967/210156311-e7d1a017-74ce-4ce7-b3a7-25e96f5baaf4.png)

Another issue is that I dont see a way to delete any tags that are added. The check mark can remove the tag from the container. Is there a way to delete a tag?

**Comment by Tryibion** [comment]
The GUI issue seems to be related to the tag not actually existing. This phantom tag was not in the Tags list in the `LayerAndTagsSettings`. It went away when I cleared the cache, but so did all of my other tags... So wouldn't really clearing the cache a fix for the bug.

**Comment by Tryibion** [comment]
There seems to be some issue with adding new tags and them being replicated to the tag settings.

![tag issue](https://user-images.githubusercontent.com/71274967/212447662-519ef6f4-64d8-4b38-bcfa-12dd6262e001.png)

**Comment by Tryibion** [comment]
@mafiesto4 I think those changes fixed the issue I was having. It got rid of the "ghost" tags and they now save much better. Thanks!

**Comment by Tryibion** [comment]
I am actually going to reopen this. Now none of the tags I make save...

**Comment by mafiesto4** [comment]
@Tryibion Do you have repro steps for that issue?

**Comment by Tryibion** [comment]
Yes, I create a tag using the new tag GUI. I can see it get added to the Tags list in the `LayersAndTagSettings` asset window. Then I just close out of the editor and reopen and the tags are gone. I am going to try it on a new project when I can and see if it is an issue with the project somehow.

**Comment by Tryibion** [comment]
Okay, I started a brand new project for 1.5. I open it up. Add some tags through the new GUI, then close the editor right after I add the tags. I open the editor back up and all the tags have disappeared.

**Comment by mafiesto4** [comment]
@Tryibion thanks for your PR! Do we need to fix/improve anything else for new tags?

**Comment by Tryibion** [comment]
@mafiesto4 yes, deleting tags. If you delete a tag in the `LayerAndTagSettings` the change will not be reflected in the picker and it will stay there until an editor restart.

**Comment by mafiesto4** [comment]
Deleting tags won't be supported because if tag was added to `Tags.List` it might be used somewhere in a project so we don't want to result in errors when loading that data - tags don't need to be added only to `LayerAndTagSettings` but can be defined in other placed too (eg. in code) - by design.

**Comment by mafiesto4** [comment]
Maybe tags not existing in `LayerAndTagSettings` (eg. added from code) could have tooltip with information about it (like `Tag added at runtime, not existing in Layers and Tags Settings.`)?

**Comment by Tryibion** [comment]
I would think that if the programmer is adding a tag from code and they don't want it in the project anymore, they will remove it from the code right? Because then it would keep being added to the tags at the moment that code is ran. I don't think being able to add them from code will effect the user experience of removing them in the editor.

**Comment by Tryibion** [comment]
If the programming is adding tags through code they are essentially managing the tags themselves, but for those who only will use the picker and never add a tag through code, the experience will be much better if they can be deleted in the editor as well.

**Comment by Tryibion** [comment]
It is an interesting problem though. Since I think where you are coming from is that let's say that an editor plugin requires certain tags to exist. They will add those tags through their initialization code. Then they will show up in the picker.... But what happens if they get deleted from the picker? Will it screw the plugin up? Would the user just need to restart the editor to get those tags back? Maybe there should be a property in `Tag` for weather it can be deleted or not? Then the code added ones have an option of setting that?

**Comment by Tryibion** [comment]
Another thought I had was what happens when the user changes out the `LayerAndTagSettings` in the game settings? I don think the `Tags` would update... It may be cool to do something there?

**Comment by Tryibion** [comment]
And another through is that since the tags are hierarchal, shouldnt the parent tag be selected as well if a child tag is selected?

**Comment by mafiesto4** [comment]
Yeah, I had similar concerns about tags but decided to keep them live even when removed from settings. Maybe in future, we could change this.

> shouldnt the parent tag be selected as well if a child tag is selected?

I wonder, maybe we should mark parent tags in the popup with `-` sign (`CheckBoxState.Intermediate`) if they are unchecked but their subtags are checked?

**Comment by Tryibion** [comment]
@mafiesto4 
1) I can see the first thing people complaining about is that they screwed up a tag name and cant delete it easily...
2) Yes, I think adding that state to parents would be good. Right now parent tags can be selected and added to a tag array along with their children (ex. Tag[Tag1, Tag1.Tag2] can both be in the array). Should this be allowed? Unreal engine's gameplay tags dont allow for this... As soon as you add a child tag, the parent is removed and only the child exists (not saying their design choice is the best, but just wondering).  Since when you check for a tag it will already check its parent tags.

**Comment by Tryibion** [comment]
Something else I was thinking about is should there be a way to use the `Tags.Get` with out creating a new tag if the tag doesn't exist? For example, using `Level.FindActor(Tags.Get("x.y.z"))` I can search for an actor with a tag using the `Tags.Get` function, but if the tag doesn't exist, I don't think it should be created since I am only looking for an actor that has a tag... Maybe there should be a bool parameter to that function to create the tag or not if not found?

**Comment by Tryibion** [comment]
Hmm I still have the issue sometimes of my tags disappearing from my `LayerAndTagsSettings`. I didnt have them open or anything this time. They just disappeared....

**Comment by mafiesto4** [comment]
@Tryibion Missing tags should be fixed in rev 969053a2407b271ac3a5fb9a7c28dfd7c1382b48
